File: 256-paint-house/256-paint-house.py
class Solution:
    def minCost(self, costs: List[List[int]]) -> int:
        n = len(costs)
        prev, curr = [0] * 3, [0] * 3
        for house in reversed(range(n)):
            for pcolor in range(3):
                curr[pcolor] = min(
                    costs[house][(pcolor + 1) % 3] + prev[(pcolor + 1) % 3],
                    costs[house][(pcolor + 2) % 3] + prev[(pcolor + 2) % 3]
                )
            prev, curr = curr, prev
        return min(prev)

File: 256-paint-house/256-paint-house.java
class Solution {
    public int minCost(int[][] costs) {
        int[] prev = new int[3], curr = new int[3];
        for(int i = costs.length - 1; i >= 0; i--){
            curr[0] = Math.min(costs[i][1] + prev[1], costs[i][2] + prev[2]);
            curr[1] = Math.min(costs[i][0] + prev[0], costs[i][2] + prev[2]);
            curr[2] = Math.min(costs[i][0] + prev[0], costs[i][1] + prev[1]);
            
            int[] tmp = prev;
            prev = curr;
            curr = tmp;
        }
        return Math.min(Math.min(prev[0], prev[1]), prev[2]);
    }
}